libtransact/src/context/manager.rs

```diff
@@ -283,6 +284,139 @@ impl<
     }
 }
 
+pub mod sync {
+    //! This module provides a thread-safe ContextManager.
+    //!
+    //! For many uses of the context manager, it will need to be shared between multiple threads,
+    //! with some threads reading and writing to a context while others create contexts.
+    use super::*;
+
+    use std::sync::{Arc, Mutex};
+
+    /// A thread-safe ContextManager.
+    #[derive(Clone)]
+    pub struct ContextManager<K, V> {
+        internal_manager: Arc<Mutex<super::ContextManager<K, V>>>,
+    }
+
+    impl<K, V> ContextManager<K, V>
+    where
+        K: Send + Hash + Eq + Clone + Debug,
+        V: Send + Clone + Debug,
+    {
+        /// Constructs a new Context Manager around a given state Read.
+        ///
+        /// The Read defines the state on which the context build.
+        pub fn new(database: Box<dyn Read<StateId = String, Key = K, Value = V>>) -> Self {
+            ContextManager {
+                internal_manager: Arc::new(Mutex::new(super::ContextManager {
+                    contexts: HashMap::new(),
+                    database,
+                })),
+            }
+        }
+
+        /// Return a set of values from a context.
+        ///
+        /// The values are returned as key-value tuples
+        ///
+        ///
+        /// # Errors
+        ///
+        /// Returns an error if the context id does not exist, or an error occurs while reading
+        /// from the underlying state.
+        pub fn get(
+            &self,
+            context_id: &ContextId,
+            keys: &[K],
+        ) -> Result<Vec<(K, Option<V>)>, ContextManagerError> {
+            self.internal_manager
+                .lock()
+                .expect("Lock in the get method was poisoned")
+                .get(context_id, keys)
+        }
+
+        /// # Errors
+        ///
+        /// Returns an error if the context id does not exist, or an error occurs while reading
+        /// from the underlying state.
+        pub fn set_state(
+            &self,
+            context_id: &ContextId,
+            key: K,
+            value: V,
+        ) -> Result<(), ContextManagerError> {
+            self.internal_manager
+                .lock()
+                .expect("Lock in set_state was poisoned")
+                .set_state(context_id, key, value)
+        }
+
+        pub fn delete_state(
+            &self,
+            context_id: &ContextId,
+            key: &K,
+        ) -> Result<Option<V>, ContextManagerError> {
+            self.internal_manager
+                .lock()
+                .expect("Lock in delete_state was poisoned")
+                .delete_state(context_id, key)
+        }
+
+        pub fn add_event(
+            &self,
+            context_id: &ContextId,
+            event: Event,
+        ) -> Result<(), ContextManagerError> {
+            self.internal_manager
+                .lock()
+                .expect("Lock in the add_event was poisoned")
```

Comment:
Half of these are "Lock in the" and half are "Lock in"

Reply:
fixed